torch/csrc/python_dimname.cpp

```diff
@@ -0,0 +1,83 @@
+#ifdef NAMEDTENSOR_ENABLED
+#include <torch/csrc/python_dimname.h>
+#include <torch/csrc/Exceptions.h>
+#include <torch/csrc/utils/python_strings.h>
+#include <c10/util/flat_hash_map.h>
+
+namespace torch {
+
+struct InternedStringsTable {
+  InternedStringsTable() = default;
+  ~InternedStringsTable();
+  InternedStringsTable(const InternedStringsTable &) = delete;
+  InternedStringsTable& operator =(InternedStringsTable const&) = delete;
+  InternedStringsTable(InternedStringsTable&&) = delete;
+  InternedStringsTable& operator=(InternedStringsTable&&) = delete;
+
+  at::optional<at::Dimname> lookup(PyObject* obj);
+  // Precondition: obj is an interned python string.
+  void addMapping(PyObject* obj, at::Dimname dimname);
+ private:
+  ska::flat_hash_map<PyObject*,at::Dimname> py_interned_string_to_dimname_;
+};
+
+InternedStringsTable kPyInternedStringToDimname;
+
+InternedStringsTable::~InternedStringsTable() {
+  for (auto it = py_interned_string_to_dimname_.begin();
+      it != py_interned_string_to_dimname_.end(); ++it) {
+    // See Note [References to python interned strings]
+    Py_DECREF(it->first);
+  }
+}
+
+at::optional<at::Dimname> InternedStringsTable::lookup(PyObject* obj) {
+  auto it = py_interned_string_to_dimname_.find(obj);
+  if (it == py_interned_string_to_dimname_.end()) {
+    return at::nullopt;
+  }
+  return it->second;
+}
+
+
+void InternedStringsTable::addMapping(PyObject* obj, at::Dimname dimname) {
+  // Note [References to python interned strings]
+  // If a Python interned string has no references to it, then it gets
+  // deallocated, invalidating this mapping. Let's immortalize the string by
+  // holding a refcount to it and releasing it in the destructor
+  Py_INCREF(obj);
+  py_interned_string_to_dimname_.emplace(obj, dimname);
+}
+
+} // namespace torch
+
+at::Dimname THPDimname_parse(PyObject* obj) {
+  if (obj == Py_None) {
+    return at::Dimname::wildcard();
+  }
+
+  if (!THPUtils_checkString(obj)) {
+    throw torch::TypeError("expected None or string for Dimname but got %s", Py_TYPE(obj)->tp_name);
+  }
+
+  if (!THPUtils_isInterned(obj)) {
+    // internStringInPlace decrefs obj and increfs the result. Because we're
+    // not actually returning the result to the user, we need to undo these.
+    // See https://docs.python.org/3/c-api/unicode.html#c.PyUnicode_InternInPlace
+    Py_INCREF(obj);
+    THPUtils_internStringInPlace(&obj);
+    Py_DECREF(obj);
+  }
+
+  auto maybeDimname = torch::kPyInternedStringToDimname.lookup(obj);
+  if (maybeDimname) {
+    return *maybeDimname;
+  }
+
+  const auto name = THPUtils_unpackString(obj);
```

Comment:
I'm not sure if I'm understanding this correctly... first we ensure the python string is interned, but then if we don't already have a dimname in python this unpack will create a new uninterned C++ string and intern that too right? 

In the implementation of `THUtils_unpackString` I see `return std::string(PyBytes_AS_STRING(obj), size);`

Do we have two full separate copies of tables of interned strings and their dim names, one in python and one in c++? Is there any way to share interned string representations between python/c++?

Reply:
> I'm not sure if I'm understanding this correctly... first we ensure the python string is interned, but then if we don't already have a dimname in python this unpack will create a new uninterned C++ string and intern that too right?

Yes.

> Do we have two full separate copies of tables of interned strings and their dim names, one in python and one in c++? Is there any way to share interned string representations between python/c++?

In C++ we have Symbols, which are interned strings in C++. In Python we have "python interned strings". Right now the table the PythonArgParser holds is used to translate between Python interned strings and C++ Symbols.

Python interned strings and Symbols are different. I don't know of a good way to unify them (also because Symbol has to be Python-agnostic since we compile a libtorch standalone...).

Reply:
hmm.. are they sharing any memory? it seems like `PyBytesObject` and `std::string` are both wrapping a `char*`.

Reply:
We can't pass around `char*` in place of `Dimname` because we need the full `Dimname` struct.

Replacing all instances of `Symbol` (in Dimname) with `char*` is a little complicated. It's hard to guarantee that a string passed in from python is the same as a c++ interned string.

Reply:
What I meant was, I'm just curious if we actually have two of the same char*'s in memory, or if the dimname->symbol->string->char* and pyobject->char* both reference the same memory

Reply:
> if the dimname->symbol->string->char* and pyobject->char* both reference the same memory

These could be different char*, or they could be the same char*. They're the same char* if the user only interacts with PyTorch via python and the name has not already been used in cpp-land.

If the name has been used in cpp-land (but not python land), and then the user creates a new name in python land, then those two names have different char*.

Reply:
ah that makes sense. thanks